tla-io/src/main/scala/at/forsyte/apalache/io/lir/PrettyWriter.scala

```diff
@@ -41,12 +43,42 @@ class PrettyWriter(
   private def prettyWriteDoc(doc: Doc): Unit = writer.write(pretty(doc, layout.textWidth).layout)
 
   def write(mod: TlaModule, extendedModuleNames: List[String] = List.empty): Unit =
-    prettyWriteDoc(toDoc(mod, extendedModuleNames))
+    prettyWriteDoc(modToDoc(mod, extendedModuleNames))
+
+  /**
+   * Pretty-prints the given decl twice: Once as valid TLA, and once (before the TLA expression) in a comment, where
+   * names of NameExs are replaced by their mapping in the NameReplacementMap. The output looks, for example, like this:
+   *
+   * {{{
+   *  (* State0 ==
+   * ♢(x = 11) = FALSE
+   * /\ ♢(x = 11)_unroll = FALSE *)
+   * State0 ==
+   * __temporal_t_1 = FALSE
+   * /\ __temporal_t_1_unroll = FALSE
+   * }}}
+   *
+   * Note that the expression and its comment are the same, but the names for {{{__temporal_t_1}}} and
+   * {{{__temporal_t_1_unroll}}} were substituted.
+   * @see
+   *   NameReplacementMap
+   */
+  def writeWithNameReplacementComment(decl: TlaDecl): Unit = {
+    val declDoc = declToDoc(decl) <> line <> line
+    if (NameReplacementMap.NameReplacementMap.isEmpty) {
+      prettyWriteDoc(declDoc)
+    } else {
+      val declComment = toCommentDoc(decl)
+      prettyWriteDoc(declComment <> line <> declDoc)
+    }
+  }
 
   // Declarations have a trailing empty line
-  def write(decl: TlaDecl): Unit = prettyWriteDoc(toDoc(decl) <> line <> line)
+  def write(decl: TlaDecl): Unit = {
+    prettyWriteDoc(declToDoc(decl) <> line <> line)
+  }
 
-  def write(expr: TlaEx): Unit = prettyWriteDoc(toDoc((0, 0), expr))
+  def write(expr: TlaEx): Unit = prettyWriteDoc(exToDoc((0, 0), expr, (x: String) => x))
```

Comment:
I think we can just use `Predef.identity` here?

```suggestion
  def write(expr: TlaEx): Unit = prettyWriteDoc(exToDoc((0, 0), expr, Predef.idenitity[String])
```

Reply:
I'm not sure I prefer it - it's longer and (to me) slightly less clear about the intent (since I'd have to convince myself identity is indeed simply the identity function).

Reply:
In this case I'd still prefer to name it, e.g.,

```suggestion
  private def identity(x: String): String = x
  def write(expr: TlaEx): Unit = prettyWriteDoc(exToDoc((0, 0), expr, identity)
```